core/src/main/java/io/grpc/CallOptions.java

```diff
@@ -52,19 +52,47 @@
    */
   public static final CallOptions DEFAULT = new CallOptions();
 
+  public static final class Key<T> {
+    private final String name;
+    private final T defaultValue;
+    
+    private Key(String name, T defaultValue) {
+      this.name = name;
+      this.defaultValue = defaultValue;
+    }
+    
+    public T getDefault() {
+      return defaultValue;
+    }
+
+    @Override
+    public String toString() {
+      return name;
+    }
+
+    /**
+     * Factory method for creating instances of {@link Key}.
+     *
+     * @param name the name of Key.
+     * @param defaultValue default value to return when value for key not set
+     * @param <T> Key type
+     * @return Key object
+     */
+    public static <T> Key<T> of(String name, T defaultValue) {
+      return new Key<T>(name, defaultValue);
+    }
+  }
+  
   // Although {@code CallOptions} is immutable, its fields are not final, so that we can initialize
   // them outside of constructor. Otherwise the constructor will have a potentially long list of
   // unnamed arguments, which is undesirable.
-  private Deadline deadline;
```

Comment:
I thought you would but the class looked kinda awkward with some fields as options and custom fields in a map.  This felt cleaner and more consistent with Attributes and Context.  If you feel strongly about this I can rework this so the fields are kept.  Let me know.